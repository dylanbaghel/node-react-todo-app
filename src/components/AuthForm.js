import React from 'react';
import classnames from 'classnames';
import isEmail from 'validator/lib/isEmail';

class AuthForm extends React.Component {
    constructor(props) {
        super(props);
        this.handleEmailChange = this.handleEmailChange.bind(this);
        this.handlePasswordChange = this.handlePasswordChange.bind(this);
        this.onFormSubmit = this.onFormSubmit.bind(this);
        this.state = {
            email: '',
            password: '',
            errors: {}
        };
    }

    handleEmailChange(e) {
        const email = e.target.value;
        this.setState(() => ({
            email
        }));
    }

    handlePasswordChange(e) {
        const password = e.target.value;
        this.setState(() => ({
            password
        }));
    }

    onFormSubmit(e) {
        e.preventDefault();

        if (!this.state.email.trim()) {
            this.setState(() => ({
                errors: { email: 'Please Enter Email' }
            }));
            return;
        } else if (!isEmail(this.state.email)) {
            this.setState(() => ({
                errors: {email: 'Please Enter Valid Email'}
            }));
            return;
        }
        if (!this.state.password.trim()) {
            this.setState(() => ({
                errors: { password: 'Please Enter Password' }
            }));
            return;
        } else if (this.state.password.length < 6) {
            this.setState(() => ({
                errors: { password: 'Password Must Be Greater Then 6 Characters.'}
            }));
            return;
        }
        this.setState(() => ({
            errors: {}
        }));
        this.props.onSubmit({
            email: this.state.email,
            password: this.state.password
        });
    }

    render() {
        return (
            <div>
                <form onSubmit={this.onFormSubmit}>
                    <div className="form-group">
                        <input
                            type="text"
                            placeholder="Email"
                            className={classnames('form-control', {
                                'is-invalid': this.state.errors.email
                            })}
                            value={this.state.email}
                            onChange={this.handleEmailChange}
                            onKeyUp={(e) => {
                                const email = e.target.value;
                                if (!isEmail(email)) {
                                    this.setState(() => ({
                                        errors: {email: 'Please Enter Valid Email'}
                                    }));
                                } else {
                                    this.setState(() => ({
                                        errors: {}
                                    }));
                                }
                            }}
                        />
                        <div className="invalid-feedback">
                            <small className="text-danger">{this.state.errors.email}</small>
                        </div>
                    </div>
                    <div className="form-group">
                        <input
                            type="password"
                            placeholder="Password"
                            className={classnames('form-control', {
                                'is-invalid': this.state.errors.password
                            })}
                            value={this.state.password}
                            onChange={this.handlePasswordChange}
                        />
                        <div className="invalid-feedback">
                            <small className="text-danger">{this.state.errors.password}</small>
                        </div>
                    </div>
                    <button className="btn btn-primary">Submit</button>
                </form>
            </div>
        );
    }
}

export default AuthForm;