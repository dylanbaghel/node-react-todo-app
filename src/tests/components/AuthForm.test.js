import React from 'react';
import { shallow } from 'enzyme';

import AuthForm from './../../components/AuthForm';

test('should render auth form component correctly', () => {
    const wrapper = shallow(<AuthForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid data', () => {
    const wrapper = shallow(<AuthForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('errors').email).toBeDefined();
    expect(wrapper).toMatchSnapshot();
});

test('should set email on change', () => {
    const value = 'a@gmail.com';
    const wrapper = shallow(<AuthForm />);
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('email')).toBe(value);
});

test('should set password on change', () => {
    const value = 'abhishek';
    const wrapper = shallow(<AuthForm />);
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('password')).toBe(value);
});

test('should call onSubmit prop for valid form submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<AuthForm onSubmit={onSubmitSpy}/>);
    wrapper.setState(() => ({
        email: 'a@gmail.com',
        password: 'abhishek'
    }));
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('errors')).toEqual({});
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        email: 'a@gmail.com',
        password: 'abhishek'
    });
});