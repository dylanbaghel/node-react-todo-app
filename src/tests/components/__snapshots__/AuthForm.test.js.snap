// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render auth form component correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="Email"
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
      >
        <small
          className="text-danger"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Password"
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
      >
        <small
          className="text-danger"
        />
      </div>
    </div>
    <button
      className="btn btn-primary"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should render error for invalid data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="Email"
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
      >
        <small
          className="text-danger"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Password"
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
      >
        <small
          className="text-danger"
        />
      </div>
    </div>
    <button
      className="btn btn-primary"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`should render error for invalid data 2`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        className="form-control is-invalid"
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="Email"
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
      >
        <small
          className="text-danger"
        >
          Please Enter Email
        </small>
      </div>
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Password"
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
      >
        <small
          className="text-danger"
        />
      </div>
    </div>
    <button
      className="btn btn-primary"
    >
      Submit
    </button>
  </form>
</div>
`;
